taqueria-plugin-tezos-client/client.ts

```diff
@@ -0,0 +1,256 @@
+    input = input.match(/^".*"$/) ? "\\" + input.slice(0, -1) + "\\\"" : input
+
+    // TODO: maybe validate storage and input value before passing it to tezos-client to prevent tezos-client menu being displayed?
+
+    const cmd = `docker exec ${sandbox.name} tezos-client run script ${sourcePath} on storage \'${storage}\' and input \'${input}\'`
+    return cmd
+}
+
+const simulateContract = (opts: Opts, sandbox: Sandbox) => async (sourceFile: string) : Promise<{contract: string, result: string}> => {
+    let sourcePath: string = getInputFilenameFromArtifactsDir(opts, sourceFile)
+    sourcePath = await execCmd(getCheckFileExistenceCommand(sandbox, sourcePath))
+                       .then(() => sourcePath)
+                       .catch(() => getInputFilenameFromContractDir(opts, sourceFile))
+
+    const simulateContractHelper = () => {
+        try {
+            const cmd = getSimulateCommand(opts, sandbox, sourceFile, sourcePath)
+            return execCmd(cmd)
+            .then(async ({stdout, stderr}) => { // How should we output warnings?
+                if (stderr.length > 0) sendErr(`\n${stderr}`)
+                return {
+                    contract: sourceFile,
+                    result: stdout
+                }
+            })
+            .catch(err => {
+                sendErr(" ")
+                sendErr(err.message.split("\n").slice(1).join("\n"))
+                return Promise.resolve({
+                    contract: sourceFile,
+                    result: "Invalid"
+                })
+            })
+        } catch (err) {
+            sendErr(" ")
+            sendErr((err as Error).message)
+            return Promise.resolve({
+                contract: sourceFile,
+                result: "Bad input or storage value"
+            })
+        }
+    }
+
+    return execCmd(getCheckFileExistenceCommand(sandbox, sourcePath))
+    .then(simulateContractHelper)
+    .catch(err => {
+        sendErr(" ")
+        sendErr(sourceFile + ": Does not exist\n")
+        return Promise.resolve({
+            contract: sourceFile,
+            result: "N/A"
+        })
+    })
+}
+
+const simulate = <T>(parsedArgs: Opts, sandbox: Sandbox): LikeAPromise<PluginResponse, Failure<T>> => {
+    if (parsedArgs.sourceFile) {
+        return simulateContract (parsedArgs, sandbox) (parsedArgs.sourceFile as string)
+               .then(data => [data])
+               .then(sendAsyncJsonRes)
+    }
+    return sendAsyncErr(`Please specify a contract. E.g. taq simulate local <contractName> ...`)
+}
+
+const simulateTask = async <T>(parsedArgs: Opts) : LikeAPromise<void, Failure<T>> => {
+    const sandbox = getSandbox(parsedArgs)
+    if (sandbox) {
+        if (doesUseFlextesa(sandbox)) {
+            return await isSandboxRunning(sandbox.name.value)
+            ? simulate(parsedArgs, sandbox).then(sendJsonRes)
+            : sendAsyncErr(`The ${sandbox.name} sandbox is not running.`)
```

Comment:
I think this is only useful to users who just started using Taqueria. After a while, it'd be quite verbose and there'd be extra syncing to be done in case we change the UI for starting a sandbox?

I don't have a strong preference.

Reply:
Agreed, and the more I think about it as well, the real solution is to have intuitive software that can detect and recover from these types of easily handled errors and that's a future issue